File: ide-extension/src/services/setup.service.ts
/**
 * Setup collection service for gathering IDE configuration
 */

import * as vscode from 'vscode';
import { UserSetup, ExtensionInfo, EditorSettings } from '../types';
import { logger } from '../utils/logger';
import { getEditorName as detectEditorName } from '../utils/ide-detector';

/** Default values for editor configuration */
const DEFAULTS = {
  THEME: 'Default Dark+',
  FONT_FAMILY: 'Consolas, Monaco, monospace',
  FONT_SIZE: 14,
} as const;

/** Settings keys to collect from IDE configuration */
const SETTINGS_KEYS = [
  // Editor settings
  'editor.tabSize',
  'editor.wordWrap',
  'editor.minimap.enabled',
  'editor.formatOnSave',
  'editor.lineHeight',
  'editor.fontLigatures',
  'editor.cursorStyle',
  'editor.cursorBlinking',
  'editor.renderWhitespace',
  'editor.rulers',
  // Workbench settings
  'workbench.iconTheme',
  'workbench.sideBar.location',
  'workbench.activityBar.location',
  // Files settings
  'files.autoSave',
  'files.trimTrailingWhitespace',
  'files.insertFinalNewline',
  // Terminal settings
  'terminal.integrated.fontSize',
  'terminal.integrated.fontFamily',
  'terminal.integrated.cursorStyle',
] as const;

/**
 * Service for collecting user's IDE setup configuration
 */
export class SetupService {
  /**
   * Get normalized editor name for API identification
   */
  getEditorName(): string {
    return detectEditorName();
  }

  /**
   * Collect complete setup data from IDE environment
   *
   * @param name - User-defined name for the setup
   * @param description - Optional description of the setup
   * @param isPublic - Whether the setup should be publicly visible
   * @returns The complete UserSetup object
   */
  async collectSetup(name: string, description?: string, isPublic = true): Promise<UserSetup> {
    logger.info('Collecting setup data...');

    const config = vscode.workspace.getConfiguration();

    const setup: UserSetup = {
      name,
      description,
      editorName: this.getEditorName(),
      theme: this.getTheme(config),
      fontFamily: this.getFontFamily(config),
      fontSize: this.getFontSize(config),
      extensions: this.getExtensions(),
      settings: this.getSettings(config),
      isPublic,
    };

    logger.info(
      `Setup collected: ${setup.extensions.length} extensions, ${Object.keys(setup.settings).length} settings`
    );
    return setup;
  }

  /**
   * Get current color theme
   */
  private getTheme(config: vscode.WorkspaceConfiguration): string {
    return config.get<string>('workbench.colorTheme') || DEFAULTS.THEME;
  }

  /**
   * Get editor font family
   */
  private getFontFamily(config: vscode.WorkspaceConfiguration): string {
    return config.get<string>('editor.fontFamily') || DEFAULTS.FONT_FAMILY;
  }

  /**
   * Get editor font size
   */
  private getFontSize(config: vscode.WorkspaceConfiguration): number {
    return config.get<number>('editor.fontSize') || DEFAULTS.FONT_SIZE;
  }

  /**
   * Get installed extensions (excluding built-in)
   */
  private getExtensions(): ExtensionInfo[] {
    const extensions = vscode.extensions.all
      .filter((ext) => !ext.packageJSON.isBuiltin)
      .map((ext) => ({
        name: ext.packageJSON.displayName || ext.packageJSON.name,
        id: ext.id,
        version: ext.packageJSON.version,
        publisher: ext.packageJSON.publisher,
        description: ext.packageJSON.description,
      }));

    logger.debug(`Found ${extensions.length} user-installed extensions`);
    return extensions;
  }

  /**
   * Get important editor and workbench settings
   */
  private getSettings(config: vscode.WorkspaceConfiguration): EditorSettings {
    const settings: Record<string, unknown> = {};

    for (const key of SETTINGS_KEYS) {
      const value = config.get(key);
      if (value !== undefined && value !== null) {
        settings[key] = value;
      }
    }

    return settings as EditorSettings;
  }
}

File: intellij-extension/src/main/kotlin/com/devkraken/setuphub/services/SetupService.kt
package com.devkraken.setuphub.services

import com.intellij.application.options.CodeStyle
import com.intellij.ide.GeneralSettings
import com.intellij.ide.plugins.PluginManager
import com.intellij.ide.ui.UISettings
import com.intellij.openapi.application.ApplicationInfo
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.colors.EditorColorsManager
import com.intellij.openapi.editor.ex.EditorSettingsExternalizable
import javax.swing.UIManager

/**
 * Service for collecting user's IDE setup configuration.
 * Gathers plugins, themes, fonts, and editor settings.
 */
@Service
class SetupService {

    companion object {
        private val LOG = Logger.getInstance(SetupService::class.java)
        private const val MAX_DESCRIPTION_LENGTH = 200
        
        fun getInstance(): SetupService = ApplicationManager.getApplication().getService(SetupService::class.java)
    }

    /**
     * Collect complete setup data from IDE environment.
     * 
     * @param name User-defined name for the setup
     * @param description Optional description
     * @param isPublic Whether the setup should be publicly visible
     * @return Complete UserSetup object
     */
    fun collectSetup(name: String, description: String?, isPublic: Boolean): UserSetup {
        LOG.info("Collecting setup: $name")
        
        return UserSetup(
            name = name,
            description = description,
            editorName = getEditorName(),
            theme = getTheme(),
            fontFamily = getFontFamily(),
            fontSize = getFontSize(),
            extensions = getExtensions(),
            settings = getSettings(),
            isPublic = isPublic
        )
    }

    /**
     * Get normalized editor name for API identification.
     * Examples: "intellij-idea", "pycharm", "webstorm", "phpstorm"
     */
    fun getEditorName(): String {
        val productName = ApplicationInfo.getInstance().versionName
        
        return productName
            .lowercase()
            .replace(" ", "-")
            .replace(Regex("[^a-z-]"), "")
            .replace(Regex("-+"), "-")
            .trim('-')
    }

    /**
     * Get display-friendly editor name (for UI purposes).
     */
    fun getEditorDisplayName(): String {
        return ApplicationInfo.getInstance().versionName
    }

    /** Uses UIManager to avoid deprecated LafManager.getCurrentLookAndFeel(); L&F is set by the IDE via UIManager. */
    private fun getTheme(): String {
        return UIManager.getLookAndFeel()?.name ?: "Unknown"
    }

    private fun getFontFamily(): String {
        return EditorColorsManager.getInstance().globalScheme.editorFontName
    }

    private fun getFontSize(): Int {
        return EditorColorsManager.getInstance().globalScheme.editorFontSize
    }

    /**
     * Get installed plugins (excluding bundled).
     * Only loaded plugins are returned; getLoadedPlugins() implies enabled/active, so we avoid deprecated PluginDescriptor.isEnabled().
     */
    private fun getExtensions(): List<ExtensionInfo> {
        return PluginManager.getLoadedPlugins()
            .filter { !it.isBundled }
            .map { plugin ->
                ExtensionInfo(
                    name = plugin.name,
                    id = plugin.pluginId.idString,
                    version = plugin.version,
                    publisher = plugin.vendor,
                    description = plugin.description?.take(MAX_DESCRIPTION_LENGTH)
                )
            }
    }

    /**
     * Collect important editor settings.
     */
    private fun getSettings(): Map<String, Any> {
        val settings = mutableMapOf<String, Any>()
        
        collectColorSchemeSettings(settings)
        collectEditorBehaviorSettings(settings)
        collectCodeStyleSettings(settings)
        collectUISettings(settings)
        collectGeneralSettings(settings)

        LOG.info("Collected ${settings.size} settings")
        return settings
    }

    private fun collectColorSchemeSettings(settings: MutableMap<String, Any>) {
        try {
            val editorScheme = EditorColorsManager.getInstance().globalScheme
            settings["editor.fontFamily"] = editorScheme.editorFontName
            settings["editor.fontSize"] = editorScheme.editorFontSize
            settings["editor.lineSpacing"] = editorScheme.lineSpacing
            settings["editor.colorScheme"] = editorScheme.name
            settings["console.fontFamily"] = editorScheme.consoleFontName
            settings["console.fontSize"] = editorScheme.consoleFontSize
            settings["console.lineSpacing"] = editorScheme.consoleLineSpacing
        } catch (e: Exception) {
            LOG.debug("Could not collect color scheme settings: ${e.message}")
        }
    }

    private fun collectEditorBehaviorSettings(settings: MutableMap<String, Any>) {
        try {
            val editorSettings = EditorSettingsExternalizable.getInstance()
            settings["editor.lineNumbers"] = editorSettings.isLineNumbersShown
            settings["editor.wordWrap"] = if (editorSettings.isUseSoftWraps) "on" else "off"
            settings["editor.caretBlinking"] = editorSettings.isBlinkCaret
            settings["editor.showWhitespace"] = editorSettings.isWhitespacesShown
            settings["editor.showIndentGuides"] = editorSettings.isIndentGuidesShown
            settings["editor.virtualSpace"] = editorSettings.isVirtualSpace
            settings["editor.breadcrumbs"] = editorSettings.isBreadcrumbsShown
            settings["editor.foldingOutline"] = editorSettings.isFoldingOutlineShown
        } catch (e: Exception) {
            LOG.debug("Could not collect editor behavior settings: ${e.message}")
        }
    }

    private fun collectCodeStyleSettings(settings: MutableMap<String, Any>) {
        try {
            val codeStyleDefaults = CodeStyle.getDefaultSettings()
            val indentOptions = codeStyleDefaults.indentOptions
            settings["editor.tabSize"] = indentOptions.TAB_SIZE
            settings["editor.indentSize"] = indentOptions.INDENT_SIZE
            settings["editor.useTabs"] = indentOptions.USE_TAB_CHARACTER
        } catch (e: Exception) {
            LOG.debug("Could not collect code style settings: ${e.message}")
        }
    }

    private fun collectUISettings(settings: MutableMap<String, Any>) {
        try {
            val uiSettings = UISettings.getInstance()
            settings["ui.showToolWindowBars"] = uiSettings.showToolWindowsNumbers
            settings["ui.compactMode"] = uiSettings.compactMode
            settings["ui.showMainToolbar"] = uiSettings.showMainToolbar
            settings["ui.showNavigationBar"] = uiSettings.showNavigationBar
            settings["ui.showStatusBar"] = uiSettings.showStatusBar
            settings["ui.wideScreenSupport"] = uiSettings.wideScreenSupport
            settings["ui.editorTabPlacement"] = getTabPlacementName(uiSettings.editorTabPlacement)
            settings["ui.recentFilesLimit"] = uiSettings.recentFilesLimit
        } catch (e: Exception) {
            LOG.debug("Could not collect UI settings: ${e.message}")
        }
    }

    private fun collectGeneralSettings(settings: MutableMap<String, Any>) {
        try {
            val generalSettings = GeneralSettings.getInstance()
            settings["general.autoSave"] = if (generalSettings.isAutoSaveIfInactive) "afterDelay" else "off"
            settings["general.autoSaveDelay"] = generalSettings.inactiveTimeout
            settings["general.reopenLastProject"] = generalSettings.isReopenLastProject
            settings["general.confirmExit"] = generalSettings.isConfirmExit
            settings["general.showTipsOnStartup"] = generalSettings.isShowTipsOnStartup
        } catch (e: Exception) {
            LOG.debug("Could not collect general settings: ${e.message}")
        }
    }

    private fun getTabPlacementName(placement: Int): String = when (placement) {
        1 -> "top"
        2 -> "left"
        3 -> "bottom"
        4 -> "right"
        0 -> "none"
        else -> "top"
    }
}
